gitpod-io/gitpod
# BillingMode: Use BillingMode to restrict API access (3/4)

## Description
This PR introduces the usage of the `getBillindMode` calculation across the code base.
It turned out quite big in terms of lines of code, but can be either reviewed commit-by-commit, or even split into multiple PRs if that increases confidence. It contains four main changes:
 - Guard calculation of BillingMode by new FF "isBillingModeEnabled": if false, getBillingMode always returns "chargebee" (in SaaS). This is meant to de-risk the rollout of "BillingModes", prior to even thinking about enabling "usage-based pricing" Feature Flag. This is meant to go away short-term (end of next week?) once we see that calling the actual impl does not break prod
 - some improvements and fixes to getBillingMode
 - Enforce that Chargebee/Stripe API is only usable with the respective BillingMode: There is one TODO: We currently guard access to the "Portal" as well, because it allows people to perform checkouts and upgrades as well, which we want to prohibit. This might block people from accessing their invoices, though. If this is a problem we should investigate in a follow-up issue.
 - Use BillingMode to guide what components are shown in the UI: here I might be wrong in some details, but the commit should convey the idea of how to use the BIllingMode for these decisions
 
### ToDos for follow-up PRs:
   - [ ] Align access to configuring WorkspaceClasses with `BillingMode.canSetWorkspaceClass`
   ~- [ ] Find a way to display "SpendingLimit" independent of "Billing" component ([here](https://github.com/gitpod-io/gitpod/blob/1e909bab0e075c2816118a001a435326bb1888f5/components/dashboard/src/settings/Billing.tsx#L14))~
   - [ ] Cleanup: Introduce a `TeamContext` and use it across dashboard
   - [ ] Allow access to old Chargebee invoices _without_ allowing them ti upgrade (Portal access)

## Related Issue(s)

Fixes #

## How to test
- `(cd components/server && leeway build .:dbtest)` : verify that all BillingMode tests are still running

### isBillingModeEnabled: false
- go to preview environment, sign up with test user 1 (U1) and see that Chargebee integration works as expected:
   1. start workspace and try to increase timeout: :x: 
   2. Upgrade to "Unleashed", start workspace and try to increase timeout: :heavy_check_mark:  
- however, you do _not_ see any usage-based UI

### isBillingModeEnabled: true
- repeat the test from above, with a new user (U2)
- notice how everything still works (now with more expensive computation, maybe)

:information_source:  Testing `isUsageBaseBilling` is actually out-of-scope for this PR

## Release Notes

```release-note
NONE
```

## Documentation


## Werft options:
